cai_causal_graph/time_series_causal_graph.py

```diff
@@ -0,0 +1,673 @@
+                            'time_lag': neglag,
+                        },
+                    )
+                    # if node does not exist, add it
+                    if not extended_graph.node_exists(lagged_node.identifier):
+                        extended_graph.add_node(node=lagged_node)
+
+            for lag in range(1, backward_steps + 1):
+                neglag = -lag
+                for node in minimal_graph.get_nodes():
+                    # add in-bound edges (here it could beyond the backward_steps as dictated by the minimal graph)
+                    for in_edge in minimal_graph.get_edges(destination=node.identifier):
+                        if in_edge.source.time_lag + neglag >= -backward_steps:
+                            # create the edge with -lag
+                            # create the new identifier from the lag
+                            lagged_source_identifier = get_name_from_lag(
+                                in_edge.source.identifier, in_edge.source.time_lag + neglag
+                            )
+                            lagged_identifier = get_name_from_lag(node.identifier, neglag)
+
+                            # check if lagged_source_identifier is in the graph and if the edge is not already there
+                            if extended_graph.node_exists(lagged_source_identifier) and not extended_graph.edge_exists(
+                                lagged_source_identifier, lagged_identifier
+                            ):
+                                lagged_node = Node(
+                                    identifier=lagged_identifier,
+                                    meta={
+                                        'variable_name': node.identifier,
+                                        'time_lag': neglag,
+                                    },
+                                )
+                                lagged_source_node = extended_graph.get_node(lagged_source_identifier)
+                                lagged_edge = Edge(
+                                    source=lagged_source_node,
+                                    destination=lagged_node,
+                                    edge_type=in_edge.get_edge_type(),
+                                )
+
+                                extended_graph.add_edge(edge=lagged_edge)
+
+            # add a waring if the backward_steps is smaller than the maximum lag in the graph
+            if backward_steps < maxlag:
+                ramaining_nodes = []
+                for node in minimal_graph.get_nodes():
+                    if abs(node.time_lag) > backward_steps:
+                        ramaining_nodes.append(node.identifier)
+                logger.warning(
+                    'backward_steps is smaller than the maximum lag in the graph, the following nodes will not be added: %s',
+                    list(set(ramaining_nodes)),
+                )
+
+        if forward_steps is not None:
```

Comment:
seems like there is a LOT of repetition here with `backward_steps`. Can we not reuse logic?